include/NiftyCounter.h

```diff
@@ -0,0 +1,60 @@
+/*
+ * NiftyCounter.h
+ *
+ * Copyright (c) 2018 Lukas W <lukaswhl/at/gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This file is licensed under the MIT license. See LICENSE.MIT.txt file in the
+ * project root for details.
+ *
+ */
+
+#pragma once
+
+/// Nifty counter, also known as "Schwarz Counter". Used for ensuring global
+/// static object initialization and destruction order.
+
+template <typename T> class _NiftyCounter_Base
+{
+public:
+	_NiftyCounter_Base()
+	{
+		if (! T::inc()) T::init();
+	}
+	_NiftyCounter_Base(const _NiftyCounter_Base& other) : _NiftyCounter_Base() {}
+
+	~_NiftyCounter_Base()
+	{
+		if (! T::dec()) T::deinit();
+	}
+};
+
+template <typename T, void C(), void D()> class _NiftyCounterCD_Base : public _NiftyCounter_Base<T>
+{
+	friend class _NiftyCounter_Base<T>;
+private:
+	static void init() { C(); }
+	static void deinit() { D(); }
+	static int inc() { return T::s_count++; }
+	static int dec() { return T::s_count--; }
```

Comment:
I think you should use the pre-decrement operator here to get the correct result.